Lib/test/test_dataclasses/__init__.py

```diff
@@ -3664,6 +3664,23 @@ class A(WithDictSlot): ...
         self.assertEqual(A().__dict__, {})
         A()
 
+    @support.cpython_only
+    def test_dataclass_slot_dict_ctype(self):
+        # https://github.com/python/cpython/issues/123935
+        from test.support import import_helper
+        # Skips test if `_testcapi` is not present:
+        _testcapi = import_helper.import_module('_testcapi')
+
+        @dataclass(slots=True)
+        class A(_testcapi.HeapCTypeWithDict):
+            __dict__: dict = {}
+        self.assertEqual(A.__slots__, ())
```

Comment:
Can you also test `__dictoffset__`? Maybe even test if before `__slots__`. Test that it's different than `-1`. Same remark for the class below.